### ISSUE TITLE
Successor #3911: isolate fresh-worker thread/determinism configuration from snapshot resumption

### ISSUE THREAD

**Issue by Unjuno** [Unjuno/agent-interface#3939]
## Predecessor and concrete residual

Preserve closed #3911 and its scientific FAIL unchanged. Its comment reports exact adapter/AdamW/cursor matches but non-bit-exact heldout logits (up to 3.82e-6) and update-only p95 above 60 ms. In exact main source `research/needle_online_lora_skill_stream_v1/runner.py`, blob `69a9266981d11eaffc7dc84349356a183b77207b`, `run_seed()` calls `torch.set_num_threads(1)` and `torch.use_deterministic_algorithms(True)`, whereas the `--resume-worker` branch calls `worker()` without either setting. This is a source finding, not yet proof of the historical failure mechanism.

Intake main b2457b746a6df06f6536585dfe2ab937aff639f4. README, docs/CURRENT_GOAL.md, ROADMAP.md, branch collection, open Issue/PR collection, recent closed Issues, #3911 source and comment were inspected via GitHub MCP. Targeted search for #3911+thread returned only #3911; broader inventories are not claimed exhaustive (large paginated collections and one transient search rate-limit error). Other workers own #3929 query-version concurrency and #3931 passive-reader crash persistence; neither is touched. An initially considered checkpoint-crash extension is deferred in favor of this direct experimental confound.

## H

H1: fresh exec workers do not inherit the parent's process-local PyTorch thread/determinism settings.
H2: with identical canonical starting checkpoint, data, initialization, optimizer and eight-update schedule, explicit 1-thread+deterministic worker configuration matches a 1-thread+deterministic in-process reference exactly. A controlled 1-vs-5-thread × deterministic-off/on factorial can discriminate thread-sensitive numerical reduction from the deterministic-algorithms flag in this environment.
Neither H implies that historical #3911 is repaired, that its 16-arrival stream passes, or that its timings generalize.

## T — smallest executed rung

Allocation `needle-resume-worker-config-20260922-01`. Additive namespace `research/system1/needle_resume_worker_config_v1/`; unique issue-specific research branch. Actual provided Linux x86_64 execution container, CPython 3.13.5, PyTorch 2.10.0+cpu, 5 reported CPUs; no Docker/Podman CLI, GPU, installation, provider, GUI or input. Probe: default torch intra/inter-op threads 5/5, deterministic false, OMP_NUM_THREADS/MKL_NUM_THREADS unset. This is NOT replication of #3911's pinned Docker/PyTorch 2.5.1 image. Direct network/DNS transport failed; repository reads/writes use GitHub MCP.

Before formal: materialize exact predecessor source with matching Git blob ID; construction only checks source/config/schema/process plumbing without optimizer training. Freeze source, harness/auditor, environment, schedule and gates before one formal orchestration.

Three fresh seeds 3912101, 3912102, 3912103. For each, build a fresh base using predecessor's exact 512-row, 400-update A pretraining and rank-2 initialization, then seal arrival-0 JSON with AdamW state. One in-process 1-thread/deterministic reference applies arrival 1 (8 exact updates) and evaluates the same 4096 heldout rows. Five fresh exec workers consume identical arrival-0 bytes and invoke the exact predecessor `main()`/`worker()` path: DEFAULT (no setters), T1_D0, T1_D1, T5_D0, T5_D1. Fixed balanced rotation of worker order by seed. Explicit settings occur before any tensor work; no predecessor source edits. Retain full states, optimizer, input/output JSON, raw logits/predictions, data/schedule digests, per-worker settings/PIDs/process exits and separately measured import/process/update durations. Total formal: 3 base trainings, 3 reference arrival updates, 15 worker arrival updates. No training seed re-use from #3911, no same-ID retry, no tuning after outcomes.

## D

H1 supported only if every DEFAULT fresh worker reports its actual default settings distinct from parent (in this probed environment 5/false vs 1/true), with source and input identities matching.
H2 supported only if all three T1_D1 workers exactly match reference adapter, optimizer, cursor, base and every float32 logit/prediction; independently reconstruct all exact comparisons from retained JSON. Separately report T1_D0/T5_D0/T5_D1 and default outcomes. Thread-mechanism support requires both one-thread arms exact and a multi-thread logit difference with unchanged predictions/state in at least one seed; otherwise retain HOLD_THREAD_MECHANISM_NOT_EXPOSED, not a forced positive. Any same-config mismatch is FAIL_SAME_CONFIG_EQUIVALENCE. Any prediction/state difference is reported explicitly and disqualifies the narrow logits-only interpretation.

Timings are descriptive: three fresh samples per arm are insufficient for a reliable tail/performance claim. Do not claim <=60 ms acceptance or a causal historical latency repair. All 15 rows, 3 references, settings, source/data/state identities and subprocess exits must independently reconcile; missing/corrupt/provenance/timeout is STOP/HOLD. Raw-only stdlib auditor must reject at least six corruption controls (missing/duplicate worker, changed logits, optimizer, settings, source identity). Formal invoked once; first result retained.

## C / U and roadmap

Synthetic A/B task only. No historical retraining/relabeling, model-provider calls, real task effects, authority, shared runtime edits, or main/old-branch cleanup. Exact JSON equality tests representation/state equality, not general cross-version reproducibility. Full 16-arrival equivalence, matched original-engine replication, historical causality, quality/latency benefit, crash durability and GUI utility remain unknown.

Roadmap: exact source reconstruction -> no-training construction -> source/gate freeze -> one three-seed factorial -> independent raw audit/corruption checks -> scoped report and additive PR -> verify main integration if checks/permissions permit. Close only this bounded successor after evidence delivery; broad ROADMAP and original FAIL remain unchanged.

**Comment by Unjuno** [comment]
Preformal freeze; formal invocation count remains 0. Branch `research/issue-3939-worker-config-20260922-v1` was created from rechecked main `b2457b746a6df06f6536585dfe2ab937aff639f4`; targeted branch search had no #3939 collision. Construction passed 10 no-training checks (exact predecessor blob, source AST, parent-only setters, exact main dispatch with no training, five fresh config probes, signed-zero bitwise audit). Final prefreeze source review added timeout stdout/stderr retention and independent freeze/inventory binding; AST checks pass. No optimizer update/model/GUI/input has occurred.

Frozen SHA-256:
- FREEZE.json: `2d4f9eba14b808da8502173987df5503bf24e94f14bf2d6cf32c921fd51f8adb`
- predecessor_runner.py: `ae12bc6b7fee156bc04ea701fa939313d0db5b5c6eb600daa2153662d82e055c` (12505 bytes, Git blob `69a9266981d11eaffc7dc84349356a183b77207b`)
- experiment.py: `29ccade5c09672549bcf1cbf0992c07fb472fa0e6c55fe83fcd3ea5aae8dfeb2`
- worker_entry.py: `f9eaeb90a85805718d3b244def27f6f98b696baeec839ac05e0ed155588ab6a8`
- audit.py: `d897419b0239faf978b454704c3c27d40e72ec736a8ea981cc448095551ca819`
- construction.py: `7982ee966cd724f72da9cb40f0b893fd6dfa9eaff685480f726a8fdee3e929c4`
- PREREGISTRATION.md: `7957829f4a45d611f40bc6715fd6347f02b1d7b8e1b5525b4914c3be9f913588`

Exact next command (once, output absent):
```
python /mnt/data/agent_research/research/system1/needle_resume_worker_config_v1/experiment.py --output /mnt/data/agent_research/research/system1/needle_resume_worker_config_v1/formal-01
```
Then the frozen stdlib raw auditor, without training:
```
python /mnt/data/agent_research/research/system1/needle_resume_worker_config_v1/audit.py /mnt/data/agent_research/research/system1/needle_resume_worker_config_v1/formal-01 --output /mnt/data/agent_research/research/system1/needle_resume_worker_config_v1/AUDIT.json
```

Preregistration precision: three seeds use cyclic order rotations 0/1/2; this is **not fully position-balanced**, so the earlier phrase 'balanced rotation' must not imply a full Latin-square timing design. Timing remains descriptive only. The predecessor envelope's original allocation string is retained solely to invoke its unmodified validator; the new wrapper allocation/seed/source receipts identify this separate experiment. No historical source/result changes and no Docker/OrbStack equivalence are claimed.

**Comment by Unjuno** [comment]
Retained first formal outcome: **STOP_EXECUTION_ENVELOPE / INCOMPLETE_DENOMINATOR**, not PASS. Source and gates were unchanged. The initial interactive tool transport returned StreamingExecNotEnabledContainerError before any shell/experiment launch (directory/stdout absent and no process, verified). The same frozen CLI was then invoked exactly once via synchronous exec. That tool hit its 120-second envelope; subsequent process inspection found no surviving experiment/worker processes. No same-ID rerun or continuation is permitted.

Retained raw contains 9/15 completed worker rows with exit 0: all five modes for seed3912101 and T1_D0/T1_D1/T5_D0/T5_D1 for seed3912102. Seed3912102 DEFAULT and all seed3912103 workers are absent. There is no COMPLETED.json or final MANIFEST.json, and no outer shell exit receipt. The frozen auditor was invoked once unchanged; it exits1 with FileNotFoundError for missing MANIFEST.json. Its stdout/stderr/exit are retained; no audit PASS is claimed.

A separate read-only diagnostic (not the frozen complete-denominator audit) found that all nine completed rows have state equality, float32-bit-exact logits and identical predictions versus their own references. Thus the historical thread-sensitive logit effect has NOT been exposed in the observed partial PyTorch2.10 data. Actual fresh defaults still differ from parent settings. This cannot satisfy the three-seed H2 or finite thread-mechanism gates; no rows are pooled into a successor.

All partial bytes, source, freeze, launch error and auditor failure will be retained additively. Next smallest execution repair is a separately registered successor with three fresh seeds and immutable per-seed batches, each serialized before returning to the tool, retaining the same five arms/400-base-updates/one-arrival/4096-row decisions. It changes execution granularity, not gates; old seeds3912101-3912103 are consumed and remain untouched.

**Comment by Unjuno** [comment]
Closing the bounded allocation with its original **STOP_EXECUTION_ENVELOPE_INCOMPLETE_DENOMINATOR** unchanged. Complete source/freeze/construction and all93 retained artifact files, including nine completed rows, timeout/launch receipts and the unchanged auditor's missing-MANIFEST traceback, are now delivered via PR #4002 merged to main at `1c817de45cf944d144d76e288823b63bc0ec73a4`. Evidence entry point: `research/system1/needle_worker_config_delivery_3939_3958/README.md`.

Successor #3958 separately completed three fresh seed batches and received HOLD_THREAD_MECHANISM_NOT_EXPOSED, not a repair of this STOP or #3911 FAIL. No old seed/data was retried or pooled. All238 members across both original archives were byte-verified after decoding; canonical STOP tar.xz remains843632bytes/SHA256 `d4b7738efc0ab2e4b7392eacae08cbeb4d70e2fcad9aaad16b80d23cfc81d3b6`.

Branch cleanup check: `research/issue-3939-worker-config-20260922-v1` still points exactly to intake main `b2457b746a6df06f6536585dfe2ab937aff639f4` and has no open PR with it as head or base. It contains no unique commits. It is a cleanup candidate, not a lost-evidence branch. The exposed GitHub MCP actions include file deletion and ref updates but no branch/ref deletion; no unsupported deletion or unrelated branch change was attempted. Publication branch retention/cleanup is separate from the scientific result.
